firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -119,7 +134,111 @@ public void onFailure(@NonNull @NotNull Exception e) {
+        && !checkForUpdateTaskCompletionSource.getTask().isComplete()) {
+      checkForUpdateCancellationSource.cancel();
+    }
+
+    checkForUpdateCancellationSource = new CancellationTokenSource();
+    checkForUpdateTaskCompletionSource =
+        new TaskCompletionSource<>(checkForUpdateCancellationSource.getToken());
+    checkForUpdateHandler = HandlerCompat.createAsync(Looper.getMainLooper());
+
+    firebaseInstallationsApi
+        .getId()
+        .addOnSuccessListener(
+            new OnSuccessListener<String>() {
+              @Override
+              public void onSuccess(String fid) {
+                firebaseInstallationsApi
+                    .getToken(true)
+                    .addOnSuccessListener(
+                        new OnSuccessListener<InstallationTokenResult>() {
+                          @Override
+                          public void onSuccess(InstallationTokenResult installationTokenResult) {
+
+                            String authToken = installationTokenResult.getToken();
+                            String appId = firebaseApp.getOptions().getApplicationId();
+                            String apiKey = firebaseApp.getOptions().getApiKey();
+
+                            // make into method parameters
+                            FirebaseAppDistributionTesterApiClient client =
+                                new FirebaseAppDistributionTesterApiClient();
+
+                            handleLatestReleaseFromClient(client, fid, appId, apiKey, authToken);
+                          }
+                        });
+              }
+            })
+        .addOnFailureListener(
+            new OnFailureListener() {
+              @Override
+              public void onFailure(@NonNull Exception e) {
+                setCheckForUpdateTaskCompletionError(
+                    new FirebaseAppDistributionException(
+                        FirebaseAppDistributionException.Status.UNKNOWN));
+              }
+            });
+
+    return checkForUpdateTaskCompletionSource.getTask();
+  }
+
+  private void handleLatestReleaseFromClient(
+      FirebaseAppDistributionTesterApiClient client,
+      String fid,
+      String appId,
+      String apiKey,
+      String authToken) {
+    checkForUpdateExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              AppDistributionRelease latestRelease =
+                  client.fetchLatestRelease(fid, appId, apiKey, authToken);
+
+              Context context = firebaseApp.getApplicationContext();
+              long currentInstalledVersionCode = getInstalledAppVersionCode(context);
+
+              if (Long.parseLong(latestRelease.getBuildVersion()) <= currentInstalledVersionCode) {
+                // already on latest release
+                latestRelease = null;
+              }
+
+              AppDistributionRelease finalLatestRelease = latestRelease;
```

Comment:
As I mentioned above, checkForUpdateHandler does not need to be a class variable. You can make that variable function scope. Since we could probably re-use this logic in the future, we can probably create a new method called `updateOnUiThread` that takes in a runnable. And you can probably call `HandlerCompat.createAsync(Looper.getMainLooper())` in that method